File: example/ios/ExampleTests/ReactNativePerformanceTests.m
//
// Copyright (c) Microsoft Corporation. All rights reserved.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.
//

#import <XCTest/XCTest.h>

#import <React/RCTBridge.h>
#import <React/RCTDevSettings.h>
#import <React/RCTLog.h>

@interface ReactNativePerformanceTests : XCTestCase <RCTBridgeDelegate>
@end

@implementation ReactNativePerformanceTests {
    RCTDevSettings *_devSettings;
    BOOL _wasDebuggingRemotely;
}

- (void)setUp
{
    // Make sure that remote debugging is disabled otherwise we'll get a RedBox.
    _devSettings = [[RCTDevSettings alloc] init];
    if (_devSettings.isDebuggingRemotely) {
        _wasDebuggingRemotely = YES;
        _devSettings.isDebuggingRemotely = NO;
    }

    RCTSetLogFunction(^(__unused RCTLogLevel level,
                        __unused RCTLogSource source,
                        __unused NSString *fileName,
                        __unused NSNumber *lineNumber,
                        __unused NSString *message){
        // noop
    });
    RCTSetFatalHandler(nil);
}

- (void)tearDown
{
    if (_wasDebuggingRemotely) {
        _devSettings.isDebuggingRemotely = YES;
    }

    RCTSetLogFunction(nil);
}

- (void)testBridgeInitializationPerformance
{
    [self measureBlock:^{
      XCTestExpectation *expectation =
          [self expectationForNotification:RCTJavaScriptDidLoadNotification object:nil handler:nil];
      RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:nil];
      [self waitForExpectations:@[expectation] timeout:5];
      [bridge invalidate];
    }];
}

// MARK: - RCTBridgeDelegate details

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
    return [NSBundle.mainBundle URLForResource:@"main" withExtension:@"jsbundle" subdirectory:nil];
}

- (NSArray<id<RCTBridgeModule>> *)extraModulesForBridge:(RCTBridge *)bridge
{
    return @[];
}

@end

File: ios/ReactTestAppTests/ReactNativePerformanceTests.swift
//
// Copyright (c) Microsoft Corporation. All rights reserved.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.
//

import XCTest

@testable import ReactTestApp

class ReactNativePerformanceTests: XCTestCase, RCTBridgeDelegate {

    private var wasDebuggingRemotely = false

    override func setUp() {
        // Make sure that remote debugging is disabled otherwise we'll get a RedBox.
        let devSettings = RCTDevSettings()
        if devSettings.isDebuggingRemotely {
            wasDebuggingRemotely = true
            devSettings.isDebuggingRemotely = false
        }
    }

    override func tearDown() {
        // Reset setting for remote debugging.
        if wasDebuggingRemotely {
            RCTDevSettings().isDebuggingRemotely = wasDebuggingRemotely
        }
    }

    func testBridgeInitializationPerformance() {
        var bridge: RCTBridge?
        measure {
            expectation(forNotification: .RCTJavaScriptDidLoad, object: nil, handler: nil)

            DispatchQueue.global(qos: .userInitiated).async {
                bridge = RCTBridge(delegate: self, launchOptions: [:])
            }

            waitForExpectations(timeout: 5, handler: nil)
        }

        guard bridge?.isValid == true else {
            XCTFail("Failed to initialize React Native bridge")
            return
        }
    }

    // MARK: - RCTBridgeDelegate details

    func sourceURL(for bridge: RCTBridge!) -> URL! {
        return Bundle(for: ReactInstance.self).url(
            forResource: "index.mobile",
            withExtension: "jsbundle",
            subdirectory: nil
        )
    }

    func extraModules(for bridge: RCTBridge!) -> [RCTBridgeModule] {
        return []
    }
}
